src/merkle_tree/constraints.rs

```diff
@@ -101,39 +45,166 @@ where
+
+        let left_hash = self
+            .leaf_is_right_child
+            .select(leaf_sibling_hash, &claimed_leaf_hash)?
+            .to_bytes()?;
+        let right_hash = self
+            .leaf_is_right_child
+            .select(&claimed_leaf_hash, leaf_sibling_hash)?
+            .to_bytes()?;
+
+        let mut curr_hash = TwoToOneH::evaluate(two_to_one_hash_params, &left_hash, &right_hash)?;
+        // To traverse up a MT, we iterate over the path from bottom to top (in reverse)
+
+        // At any given bit, the bit being 0 indicates our currently hashed value is the left,
+        // and the bit being 1 indicates our currently hashed value is on the right.
+        // Thus `left_hash` is the sibling if bit is 1, and it's the computed hash if bit is 0
+        for (bit, sibling) in self.path.iter().rev().zip(self.auth_path.iter().rev()) {
+            let left_hash = bit.select(sibling, &curr_hash)?;
+            let right_hash = bit.select(&curr_hash, sibling)?;
+
+            curr_hash = TwoToOneH::evaluate(
+                two_to_one_hash_params,
+                &left_hash.to_bytes()?,
+                &right_hash.to_bytes()?,
+            )?;
+        }
+
+        Ok(curr_hash)
+    }
+
+    /// Given a leaf, calculate the hash of the merkle tree
+    /// along the path and check if the given root is correct.
+    pub fn verify_membership(
+        &self,
+        leaf_hash_params: &LeafH::ParametersVar,
+        two_to_one_hash_params: &TwoToOneH::ParametersVar,
+        root: &TwoToOneH::OutputVar,
+        leaf: &impl ToBytesGadget<ConstraintF>,
+    ) -> Result<Boolean<ConstraintF>, SynthesisError> {
+        let expected_root = self.calculate_root(leaf_hash_params, two_to_one_hash_params, leaf)?;
+        // enforce `expected_root` is equal to the given root
+        Ok(expected_root.is_eq(root)?)
+    }
+
+    /// update the `old_leaf` to `new_leaf` and returned the updated MT root. `self` will not be
+    /// modified.
+    ///
+    /// If the `old_leaf` does not lead to `old_root`, constraints will not be satisfied.
+    pub fn update_leaf(
+        &self,
+        leaf_hash_params: &LeafH::ParametersVar,
+        two_to_one_hash_params: &TwoToOneH::ParametersVar,
+        old_root: &TwoToOneH::OutputVar,
+        old_leaf: &impl ToBytesGadget<ConstraintF>,
+        new_leaf: &impl ToBytesGadget<ConstraintF>,
+    ) -> Result<TwoToOneH::OutputVar, SynthesisError> {
+        self.verify_membership(leaf_hash_params, two_to_one_hash_params, old_root, old_leaf)?
+            .enforce_equal(&Boolean::TRUE);
+        Ok(self.calculate_root(leaf_hash_params, two_to_one_hash_params, new_leaf)?)
+    }
+
+    /// update the `old_leaf` to `new_leaf` and returned the updated MT root.`self` will not be modified.
+    ///
+    /// If the `old_leaf` does not lead to `old_root`, or `new_leaf` does not lead to `new_root`,
+    /// then constraints will not be satisfied.
+    pub fn update_and_check(
```

Comment:
this should again return a bit instead of enforcing equality

Reply:
fixed